File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* helper(vector<int>& preorder, int &idx, int upper_bound) {
        int n = preorder.size();
        if(idx >= n) {
            return NULL;
        }
        if(preorder[idx] >= upper_bound) {
            return NULL;
        }
        int root_value = preorder[idx];
        TreeNode* root = new TreeNode(root_value);
        idx++;
        root -> left = helper(preorder, idx, root_value);
        root -> right = helper(preorder, idx, upper_bound);
        return root;
    }
    
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        int upper_bound = 2000, idx = 0;
        return helper(preorder, idx, upper_bound);
    }
};

File: 1008-construct-binary-search-tree-from-preorder-traversal/1008-construct-binary-search-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public int idx=0;
    
    public TreeNode helper(int[] preorder,int upperBound) {
        int n=preorder.length;
        if(idx>=n || preorder[idx]>=upperBound) {
            return null;
        }
        TreeNode node=new TreeNode(preorder[idx++]);
        node.left=helper(preorder,node.val);
        node.right=helper(preorder,upperBound);
        return node;
    }
    
    public TreeNode bstFromPreorder(int[] preorder) {
        int n=preorder.length;
        return helper(preorder,1005);
    }
}